File: Data Structures & Algorithms/search-2d-matrix/submission-1.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int ROWS = matrix.length;
        int COLS = matrix[0].length;

        int left = 0, right = ROWS * COLS - 1;

        while (left <= right){
            int middle = left + (right - left) / 2;
            int row = middle / COLS, col = middle % COLS;
            if (matrix[row][col] == target){
                return true;
            }else if (matrix[row][col] > target){
                right = middle - 1;
            }
            else{
                left = middle + 1;
            }
        }
        return false;
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-1.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        directions = [[1,0],[-1,0],[0,-1],[0,1],[0,0]]
        area = 0
        ROWS, COLS = len(grid),len(grid[0])
        def bfs(r,c):
            que = deque()
            que.append((r,c))
            islandSize = 0
            while que:
                row,col = que.popleft()
                for dr,dc in directions:
                    newRow, newCol = row+dr, col+dc
                    if newRow < 0 or newCol < 0 or newRow >= ROWS or newCol >= COLS or grid[newRow][newCol] == 0:
                        continue
                    que.append((newRow,newCol))
                    islandSize+=1
                    grid[newRow][newCol] = 0
            return islandSize
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 1:
                    area = max(area,bfs(r,c))
        return area

